Add CLion IDE developer documentation

JanVogelsang opened the pull request:
This PR adds developer documentation on how to setup the CLion IDE for NEST, which is a popular commercial IDE for C++ projects.

[edit] review docs [here](https://nest-simulator--3045.org.readthedocs.build/en/3045/developer_space/workflows/nest_with_ides.html#clion)

jessica-mitchell commented on doc/htmldoc/developer_space/workflows/nest_with_ides.rst:
```diff
@@ -410,7 +410,42 @@ Code Completion in Xcode
+
+.. image:: ../../static/img/clion_cmake_config.png
+
+3. Add the following CMake option: ``-DPython_ROOT_DIR=[path/to/root/of/python/env]``
```
Is this the only thing you need to add to the cmake config?

It is also unclear to me how you navigate from the cmake configuration to the run configuration based on the instructions and screenshots

jessica-mitchell commented on doc/htmldoc/developer_space/workflows/nest_with_ides.rst:
```diff
@@ -410,7 +410,42 @@ Code Completion in Xcode
+
+.. image:: ../../static/img/clion_run_configuration.png
+
+5. Add a ``custom build application``
```
Again, unclear how you did this

jessica-mitchell reviewed (changes requested):
thanks for adding this. However, I find the instructions not clear enough and I don't think all these screen shots are necessary. i would suggest being more explicit in the text and only keeping the run configuration screen shot and maybe the cmake one.
For example
" In the left panel select Build,Execution,Deployment -->CMake. Then add CMake options . . ."
" Navigate to the ??? to edit the run configuration.  "
Then explain which items need to be edited (e.g., "Change the working directory to the path to the source directory for NEST" ).

JanVogelsang commented on doc/htmldoc/developer_space/workflows/nest_with_ides.rst:
```diff
@@ -410,7 +410,42 @@ Code Completion in Xcode
+5. Now Xcode generates its index and after that code completion should work.
+
+
+CLion
```
You mean a link to the IDE product page? I.e., https://www.jetbrains.com/clion/ ?

JanVogelsang commented:
Alright, I addressed your comments and made the instructions clearer. @otcathatsya Could you have a look at this as well and tell me if these instructions are clear enough?

JanVogelsang commented:
It doesn't look very pretty yet, for some reason all lines that are followed by an unordered list are bold. I couldn't find out how to fix that, could you do that Jessica?

heplesser reviewed (changes requested):
Some of the instructions are horribly outdated, e.g., mention OpenMPI 1.6 and MacPorts, which hardly anyone uses any more. I will try to take a look at the Mac part.

otcathatsya commented:
> Alright, I addressed your comments and made the instructions clearer. @otcathatsya Could you have a look at this as well and tell me if these instructions are clear enough?

I think to a dev using CLion it's pretty clear, though maybe important to state that most of the steps described are to be able to properly debug NEST with the built-in IDE tools, since the "regular" project will work pretty much out of the box from my experience.

JanVogelsang commented:
> Some of the instructions are horribly outdated, e.g., mention OpenMPI 1.6 and MacPorts, which hardly anyone uses any more. I will try to take a look at the Mac part.

Alright, please open a separate issue (or ideally also a PR) for that, as this is unrelated to the documentation for using CLion.

JanVogelsang commented:
> > Alright, I addressed your comments and made the instructions clearer. @otcathatsya Could you have a look at this as well and tell me if these instructions are clear enough?
> 
> I think to a dev using CLion it's pretty clear, though maybe important to state that most of the steps described are to be able to properly debug NEST with the built-in IDE tools, since the "regular" project will work pretty much out of the box from my experience.

Right, that's true. The run configuration is required to debug C++ with CLion and the CMake configuration is fully optional, but helps with two things: 1. If not properly setup (e.g. just leaving the defaults), CLion reruns CMake from time to time (I think when checking out a new branch for example) and thus overwrites the files one generated when manually running CMake from the command line. 2. Even though I personally don't do that, one does not need the terminal at all anymore, as `cmake`, `make`, `make install`, and `python ..` are all run from CLion itself now.  
I'll add that to the documentation.

terhorstd commented:
Note that I added a link to review the rendered documentation into the PR description.

JanVogelsang commented:
> Note that I added a link to review the rendered documentation into the PR description.

Will this be automatically rebuild now that I made some more changes?

jessica-mitchell commented:
@JanVogelsang FYI: here is the link to the Read the Docs build: https://nest-simulator--3045.org.readthedocs.build/en/3045/developer_space/workflows/nest_with_ides.html#clion

jessica-mitchell commented:
> It doesn't look very pretty yet, for some reason all lines that are followed by an unordered list are bold. I couldn't find out how to fix that, could you do that Jessica?

This is because you need new lines around your second order bullet lists. Also in the first paragraph, you are missing a new line before the bullet list so the bullets are ignored.

Good rule of thumb in reStructured text is always include new lines around markup for headings, lists, directives etc.

JanVogelsang commented:
Alright, thanks, it looks good now.

heplesser commented on doc/htmldoc/developer_space/workflows/nest_with_ides.rst:
```diff
@@ -410,7 +410,48 @@ Code Completion in Xcode
+* CLion's default CMake configuration creates the build directory inside the source code root directory, which easily leads to unknowingly committing build files to git.
+* CLion regenerates build files when checking out a new branch, potentially overriding the build you created by manually running CMake from the command line.
+* After setting up both the CMake configuaration and a run configuration, you won't need to use the command line to run ``cmake``, ``make``, ``make install``, and ``python *.py`` at all anymore. All will be done automatically by CLion and you only need to run your application without having to manually recompile.
+
```
Please add this important bit of information:
```suggestion

.. note::

   CLion is a commercial product. It is *your responsibility* to ensure that you have a valid
   license permitting you to use CLion (or any software product) for your work on the    
   NEST Simulator. 
```

JanVogelsang replied:
Agreed, this is indeed an important note for all commercial products!

heplesser reviewed (changes requested):
@JanVogelsang I see your point about not fixing outdated, unrelated material. I will look at that in a separate PR. Please see the suggestion I added on CLion.

jessica-mitchell commented:
@JanVogelsang I wanted to create a PR against your branch because I had a couple of changes I was trying out myself. But your repo isn't listed in the base repositories. This happens quite frequently, and I thought it might be because of private repos but yours is public. Do you happen to know why this is?
These are the [changes](https://github.com/jessica-mitchell/nest-simulator/commit/c6b1d65af4a85a8985b85047656d83a9aa87ba71) I wanted to make. Could you apply the commit to your branch?

JanVogelsang commented:
@jessica-mitchell No idea why that is the case, but sure, I can apply the commit.

jessica-mitchell commented on doc/htmldoc/developer_space/workflows/nest_with_ides.rst:
I don't know how, but it seems like not all the changes that I did are present here.

JanVogelsang replied:
Which change are you missing? I thought I added everything..

jessica-mitchell commented on doc/htmldoc/developer_space/workflows/nest_with_ides.rst:
```diff
@@ -409,7 +409,53 @@ Code Completion in Xcode
+~~~~~~~~~~~~~~~~~~~~~~
+
+1. :ref:`Clone NEST <git_workflow>` onto your computer or select ``Get from VCS`` from CLion's start window
+2. You will be prompted to setup CMake. You can find the CMake configuration in the settings (Build,Execution,Deployment --> CMake) to make more changes at a later point. Make the following changes and press "Ok":
```
```suggestion
2. You will be prompted to setup CMake. You can find the CMake configuration in the settings (Go to :menuselection:`Build,Execution,Deployment --> CMake`) to make more changes at a later point. 

Make the following changes and press :menuselection:`Ok`:
```

JanVogelsang replied:
Done

JanVogelsang commented on doc/htmldoc/api_function_list.json:
Done
